man/slp.8
.TH "slp" 8
.SH NAME
slp \- Cobalt service location protocol
.SH SYNOPSIS
.B slp
.I [-d] [--config-files <configfile1:configfile2>] [-D=<pidfile>]
.SH "DESCRIPTION"
.PP
.B slp
Service location protocol
.SH "OPTIONS"
.TP
.B \-d
Turn on communication debugging.
.TP
.B \-\-config-files <configfile1:configfile2>
Specify path to cobalt.conf.
.TP
.B \-D=<pidfile>
Daemonize, and write pid to pidfile.
.SH "CONFIGURATION FILE PARAMETERS"
.TP
.B [components]
.TP
.B service-location
Host and port of node running the slp component in the form https://servicenode:port
.TP
.B [communication]
.TP
.B password
Shared password for components
.PP
.B key
Path to SSL key
.SH "SEE ALSO"
.BR slpstat (1)


CHANGES
= Changes from previous Cobalt Versions =

== Changes to 0.99.29 ==
 * Interactive jobs are experimentally available for cluster systems.

== Changes to 0.99.28 ==
 * Cobalt now supports qsub options in script mode jobs similarly to other
   job scheduler scripts.  The form is #Cobalt <qsub option> <option argument>.
   Most qsub options are supported, except for mode.  Jobs using this feature must
   be script mode jobs.  This is similar in style to PBS-directives. (Cobalt Ticket 354)
 * Fixed a bug in cluster systems where a failure duing job startup could result
   in all nodes for a job being stuck in the allocated state (Cobalt Ticket 662)
 * $COBALT_STARTTIME and $COBALT_ENDTIME added to job's environments, these are
   the Unix timestamp for when the job started, and when Cobalt will terminate
   the job due to exceeding walltime. (Cobat Ticket 593)
 * A requested walltime of 0 will set the job to run for the time remaining in
   a reservation. (Cobalt Ticket 474)
 * The --env flag to qsub now aggregates multiple instances of the flag rather
   than overriding the prior instance (Cobalt Ticket 657)

== Changes to 0.99.27 ==
 * Alternate kernel support for BG/Q compute nodes and IO nodes is available.
  Numerous configuration settings have been added to support this.
  (Cobalt ticket 621)
 * If $jobid is included in a job name specification or an environment variable,
  at submission time, Cobalt will substitute the generated jobid for that job.
  See man qsub for more details. (Cobalt ticket 617)
 * The Cobalt jobid is now explicitly added in the cobaltlog file. (Cobalt ticket 390)
 * The name of a job can now be specified seperately from naming the output
  files. (Cobalt ticket 625)
 * Multiple locations can be passed to reservations through the -p option as a
  ':'-delimited list. (Cobalt ticket 405)
 * cqadm can now be used to release user hold, as well as put jobs into a user hold
  as well as an admin hold (Cobalt ticket 382)
 * The users field can be removed from a reservation using setres -m -u '*'. (Cobalt
   ticket 381)
 * Releaseres should now report the correct partiton count on reservation release
  (Cobalt ticket 470)
 * Reservations now show time remaining in showres once the reservation is active.
  (Cobalt ticket 585)
 * Schedctl should handle flipping score and jobid arguments more gracefully and 
  no longer silently discards invalid options. (Cobalt ticket 502)
 * ':' and '=' may now be escaped as '\:' and '\=' in the --envs argument (Cobalt
  ticket 589)
 * Reservations may now take 'now' for the start time of a reservation (Cobalt
  ticket 546)
 * "GEOMETRY", "ION_KERNEL" and "ION_KERNELOPTIONS have been added to the JOB_DATA
  table of the Cobalt database 


== Changes to 0.99.26 ==
 * Corrected a bug that was causing messages in the clients that used to go to
  stdout to go to stderr instead

== Changes to 0.99.25 ==
 * Fixes for clients that were causing erronious default values for proccount
  have been fixed.
 * The setgid wrapper now calls a site-selectable python interperter with the
  '-E' flag by default.
 * A bug that could cause TimeRemaining to show negative values has been fixed
 * Queues may be added or removed from blocks without overwriting the entire
  list with new options in partadm using --rmq or --appq with the --queue option.


== Changes to 0.99.24 ==
 * Reverting client changes.  Found some issues when deploying to test systems
 * --defer flag has been reverted with these as a casualty.

== Changes to 0.99.23 ==
 * Client code has undergone a refactor.  Numerous bugs have been addressed
  with these that were found while improving test coverage of this code
 * IO Block booting and automatic control has been added.  Please consult the
  partadm manpage for details
 * This release requires pybgsched-0.1.3
 * A user may voluntarily set their job's score to 0 and let it re-accrue using
  qalter --defer (Ticket #590)

== Changes to 0.99.22 ==
 * Fix for a deadlock that could occur due to a race condition in the 
  boot-reaping code.
 * Fix for a pid collision that could occur after a forced delete of a job in
  the system-script-forker component leading to a job getting stuck in the
  Job_Epilogue state.
 * python 2.7 and it's change to the XMLRPC libraries are now supported.

== Changes to 0.99.21 ==
 * Minor fix release to correct a failure mode where we can lose a process group
  if a job's boot is delayed.

== Changes to 0.99.20 ==
 * Small changeset to fix a race condition that was hanging jobs that came in
   on 0.99.19
 * Better error logging for situations where the boots and the process group
  information haven't lined up yet.  This usually happens during high-load and
  can result in block boot completion never being detected.

== Changes to 0.99.19 ==
 * Manual boot control has been added
 * major refactor of compute block rebooting

== Changes to 0.99.18 ==
 * Fix for other systems that were seeing erronious geometry data
 * SoftwareFailure now accurately reported

== Changes to 0.99.17 ==

 * Support for --geometry has been added.  This will constrain a job to run 
blocks having a specific geometry in nodes, partadm, partlist and qstat have
been updated to show this.
 * Support for passthrough blocking reservations has been added.
 * Fixed a bug where a job run outside of Cobalt would not properly block resources.
 * Removed autoreboot in the event of a block entering a control system 
  SoftwareFalure state on BG/Q.  Normal cleanup should handle this now.

== Changes to 0.99.15 ==
 * Fix for an issue that was causing partadm to malfunction when reservations
were set on the system.

== Changes to 0.99.12 ==
 * As a note, this is a release of fixes and backported features from BG/Q
 * Fixed issues that could cause a bad hardware state to be overridden as 
"blocked" allowing the block to be drained
 * Fix for a situation where a partition that is inelligible to run due to a
pending reservation is selected for draining.


src/lib/Components/slp.py
"""Implementations of the service-location component.

Classes:
ServiceLocator -- generic implementation
PollingServiceLocator -- network-aware
TimingServiceLocator -- timeout-based

The service-location component provides registration and lookup functions
to store the locations of dynamically addressed xmlrpc servers. It is used
directly by ComponentProxy to connect to components whose service location
is not configured statically locally.

This module currently provides three implementations of service-location:

The generic implementation, ServiceLocator, is completely passive, and
has no logic regarding the expiration or validity of the services and
locations registered. It is little more than a hash, storing service data
verbatim as it is registered and unregistered.

PollingServiceLocator extends the generic implementation by automatically
polling all registered services by calling their "ping" methods. This
verifies that (a) the component is serving at the specified location, and
(b) the component is responding.

TimingServiceLocator extends the generic implementation by aotomatically
expiring any service that has not registered in a given timeframe (specified
at initialization with the "expire" keyword argument). Services are expected
to register themselves regularly at an interval <= that timeframe.
"""

import logging
import sys
import socket
import time
from xmlrpclib import ServerProxy

import Cobalt.Logging
from Cobalt.Data import Data, DataDict
from Cobalt.Components.base import Component, exposed, automatic, query
from Cobalt.Server import XMLRPCServer


__all__ = [
    "ServiceLocator", "PollingServiceLocator", "TimingServiceLocator",
]


class Service (Data):
    
    fields = Data.fields + ["tag", "name", "location"]
    
    def __init__ (self, spec):
        Data.__init__(self, spec)
        spec = spec.copy()
        self.name = spec.pop("name")
        self.location = spec.pop("location")
        self.tag = spec.get("tag", "service")

        self.stamp = time.time()

    def touch(self):
        self.stamp = time.time()


class ServiceDict (DataDict):
    
    item_cls = Service
    key = "name"


class ServiceLocator (Component):
    
    """Generic implementation of the service-location component.
    
    Methods:
    register -- register a service (exposed)
    unregister -- remove a service from the registry (exposed)
    locate -- retrieve the location of a service (exposed)
    get_services -- part of the query interface from DataSet (exposed)
    """
    
    name = "service-location"
    
    # A default logger for the class is placed here.
    # Assigning an instance-level logger is supported,
    # and expected in the case of multiple instances.
    logger = logging.getLogger("Cobalt.Components.ServiceLocator")
    
    def __init__ (self, *args, **kwargs):
        """Initialize a new ServiceLocator.
        
        All arguments are passed to the component constructor.
        """
        Component.__init__(self, *args, **kwargs)
        self.services = ServiceDict()
    
    def __getstate__(self):
        state = {}
        state.update(Component.__getstate__(self))
        state.update({
                'slp_version': 1})
        return state
   
    def __setstate__(self, state):
        Component.__setstate__(self, state)

    def register (self, service_name, location):
        """Register the availability of a service.
        
        Arguments:
        service_name -- name of the service to register
        location -- location of the service
        """
        try:
            service = self.services[service_name]
        except KeyError:
            service = Service(dict(name=service_name, location=location))
            self.services[service_name] = service
            self.logger.info("register(%r, %r)" % (service_name, location))
        else:
            service.location = location
            service.touch()
    register = exposed(register)
    
    def unregister (self, service_name):
        """Remove a service from the registry.
        
        Arguments:
        service_name -- name of the service to remove
        """
        try:
            del self.services[service_name]
        except KeyError:
            self.logger.info("unregister(%r) [not registered]" % (service_name))
        else:
            self.logger.info("unregister(%r)" % (service_name))
    unregister = exposed(unregister)
    
    def locate (self, service_name):
        """Retrieve the location for a service.
        
        Arguments:
        service_name -- name of the service to look up
        """
        try:
            service = self.services[service_name]
        except KeyError:
            self.logger.debug("locate(%r) [not registered]" % (service_name))
            return ""
        return service.location
    locate = exposed(locate)
    
    def get_services (self, specs):
        """Query interface "Get" method."""
        return self.services.q_get(specs)
    get_services = exposed(query(get_services))


class PollingServiceLocator (ServiceLocator):
    
    """ServiceLocator with active expiration.
    
    Methods:
    check_services -- ping services (automatic)
    """
    
    implementation = "polling"
    
    logger = logging.getLogger("Cobalt.Components.PollingServiceLocator")
    
    def check_services (self):
        """Ping each service to check its availability.
        
        Unregister unresponsive services.
        """
        for service in self.services.values():
            try:
                ServerProxy(service.location).ping()
            except socket.error, e:
                self.logger.warn("unable to contact %s [%s]" % (service.name, e))
                self.unregister(service.name)
            except Exception, e:
                self.logger.error("error in %s (%s)" % (service.name, e))
                self.unregister(service.name)
    check_services = automatic(check_services)


class TimingServiceLocator (ServiceLocator):
    
    """ServiceLocator with passive expiration.
    
    Attributes:
    expire -- number of seconds to expire a service
    
    Methods:
    expire_services -- check service timestamps (automatic)
    """
    
    implementation = "slp"
    
    logger = logging.getLogger("Cobalt.Components.TimingServiceLocator")
    
    def __init__ (self, expire=180, *args, **kwargs):
        """Initialize a TimingServiceLocator.
        
        Keyword arguments:
        expire -- Number of seconds when services expire.
        
        Additional arguments are passed to ServiceLocator.
        """
        ServiceLocator.__init__(self, *args, **kwargs)
        self.expire = expire
    
    def expire_services (self):
        """Check each service timestamp.
        
        Unregister expired services.
        """
        now = time.time()
        for service in self.services.values():
            if now - service.stamp > self.expire:
                self.logger.info("%s expired" % (service.name))
                self.unregister(service.name)
    expire_services = automatic(expire_services)
